File: tauri-rust/src-tauri/src/notes.rs
//! The `.md-boss` store on disk: read (three shapes fold into one), write (canonical shape,
//! atomic, removed when empty). The rules about *which* store a note belongs in, shifting
//! and deduplication live on the TypeScript side (src/models/notes.ts, annotationStore.ts);
//! this file is the IO.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;
use std::fs;
use std::io;
use std::path::Path;

/// A marked line, with or without something written about it. Paths are stored
/// tilde-abbreviated; line numbers are 1-based. Empty title/body are left out of the file,
/// so a plain jump point stays a two-key object in a file meant to be read by a person.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct Note {
    pub path: String,
    pub line: u64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub title: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub body: String,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Default)]
pub struct AnnotationFile {
    pub notes: Vec<Note>,
}

impl AnnotationFile {
    pub fn is_empty(&self) -> bool {
        self.notes.is_empty()
    }
}

/// Three shapes fold into one array: `notes`, plus `bookmarks` and `comments` written by
/// older builds. A line carrying both an old bookmark and an old comment becomes a single
/// note with a title and a body. Missing fields read as empty; a malformed or missing file
/// is empty. Never an error: a store that cannot be read is a store with nothing in it.
pub fn read_notes(path: &Path) -> AnnotationFile {
    let text = match fs::read_to_string(path) {
        Ok(text) => text,
        Err(_) => return AnnotationFile::default(),
    };
    parse(&text)
}

pub fn parse(text: &str) -> AnnotationFile {
    let raw: Value = match serde_json::from_str(text) {
        Ok(value) => value,
        Err(_) => return AnnotationFile::default(),
    };
    let obj = match raw.as_object() {
        Some(obj) => obj,
        None => return AnnotationFile::default(),
    };
    let mut found = Vec::new();
    for key in ["notes", "bookmarks", "comments"] {
        if let Some(Value::Array(items)) = obj.get(key) {
            for item in items {
                let (path, line) = match (item.get("path").and_then(Value::as_str), item.get("line").and_then(Value::as_u64)) {
                    (Some(path), Some(line)) => (path, line),
                    _ => continue,
                };
                found.push(Note {
                    path: path.to_string(),
                    line,
                    title: item.get("title").and_then(Value::as_str).unwrap_or("").to_string(),
                    body: item.get("body").and_then(Value::as_str).unwrap_or("").to_string(),
                });
            }
        }
    }
    AnnotationFile { notes: fold(found) }
}

/// One note per (path, line), first non-empty value winning per field, in path then line
/// order - the same fold as the TypeScript side, so the two agree about what a file says.
pub fn fold(notes: Vec<Note>) -> Vec<Note> {
    let mut merged: BTreeMap<(String, u64), Note> = BTreeMap::new();
    for note in notes {
        let key = (note.path.clone(), note.line);
        match merged.get_mut(&key) {
            None => {
                merged.insert(key, note);
            }
            Some(existing) => {
                if existing.title.is_empty() {
                    existing.title = note.title;
                }
                if existing.body.is_empty() {
                    existing.body = note.body;
                }
            }
        }
    }
    merged.into_values().collect()
}

/// Pretty, sorted keys, trailing newline - a file meant to be read, hand-edited and diffed.
/// Only the current key is written, so a file converts itself the first time anything in it
/// is touched.
pub fn serialize(file: &AnnotationFile) -> String {
    // serde_json without preserve_order writes maps in key order; going through Value
    // makes that the case for the structs as well.
    let value = serde_json::to_value(file).unwrap_or(Value::Null);
    let mut out = serde_json::to_string_pretty(&value).unwrap_or_else(|_| "{\"notes\":[]}".to_string());
    out.push('\n');
    out
}

/// An emptied file is removed rather than left as `{}` littering the project root. The
/// write is atomic - a temp file renamed into place - so a watcher or a git status never
/// sees half a file.
pub fn write_notes(path: &Path, file: &AnnotationFile) -> io::Result<()> {
    if file.is_empty() {
        return match fs::remove_file(path) {
            Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(()),
            other => other,
        };
    }
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let tmp = path.with_extension("md-boss.tmp");
    fs::write(&tmp, serialize(file))?;
    fs::rename(&tmp, path)
}

#[tauri::command]
pub async fn read_notes_cmd(path: String) -> Result<AnnotationFile, String> {
    tauri::async_runtime::spawn_blocking(move || read_notes(Path::new(&path)))
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn write_notes_cmd(path: String, file: AnnotationFile) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || write_notes(Path::new(&path), &file).map_err(|e| e.to_string()))
        .await
        .map_err(|e| e.to_string())?
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicUsize, Ordering};

    static COUNTER: AtomicUsize = AtomicUsize::new(0);

    fn scratch() -> std::path::PathBuf {
        let n = COUNTER.fetch_add(1, Ordering::SeqCst);
        let dir = std::env::temp_dir().join(format!("md-boss-notes-{}-{}", std::process::id(), n));
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    fn note(path: &str, line: u64, title: &str, body: &str) -> Note {
        Note { path: path.into(), line, title: title.into(), body: body.into() }
    }

    #[test]
    fn legacy_bookmarks_and_comments_fold_into_notes() {
        let text = r#"{
          "bookmarks": [{"path": "~/a.md", "line": 3, "title": "Plan"}],
          "comments": [{"path": "~/a.md", "line": 3, "body": "revisit"}, {"path": "~/b.md", "line": 1, "body": "x"}],
          "notes": [{"path": "~/a.md", "line": 9}]
        }"#;
        let file = parse(text);
        assert_eq!(
            file.notes,
            vec![note("~/a.md", 3, "Plan", "revisit"), note("~/a.md", 9, "", ""), note("~/b.md", 1, "", "x")]
        );
    }

    #[test]
    fn malformed_or_partial_records_read_as_empty_or_are_skipped() {
        assert!(parse("not json").is_empty());
        assert!(parse("[1,2]").is_empty());
        let file = parse(r#"{"notes": [{"path": "~/a.md"}, {"line": 2}, {"path": "~/c.md", "line": 2, "title": 5}]}"#);
        assert_eq!(file.notes, vec![note("~/c.md", 2, "", "")]);
    }

    #[test]
    fn canonical_shape_on_write_sorted_keys_no_empty_fields() {
        let file = AnnotationFile { notes: vec![note("~/a.md", 3, "Plan", ""), note("~/a.md", 1, "", "body")] };
        let text = serialize(&file);
        assert_eq!(
            text,
            "{\n  \"notes\": [\n    {\n      \"line\": 3,\n      \"path\": \"~/a.md\",\n      \"title\": \"Plan\"\n    },\n    {\n      \"body\": \"body\",\n      \"line\": 1,\n      \"path\": \"~/a.md\"\n    }\n  ]\n}\n"
        );
        assert!(!text.contains("bookmarks"));
    }

    #[test]
    fn round_trip_and_removal_when_emptied() {
        let dir = scratch();
        let store = dir.join(".md-boss");
        assert!(read_notes(&store).is_empty());
        let file = AnnotationFile { notes: vec![note("~/a.md", 3, "Plan", "revisit")] };
        write_notes(&store, &file).unwrap();
        assert_eq!(read_notes(&store), file);
        assert!(!dir.join(".md-boss.tmp").exists());
        write_notes(&store, &AnnotationFile::default()).unwrap();
        assert!(!store.exists());
        // removing what is already gone is fine
        write_notes(&store, &AnnotationFile::default()).unwrap();
    }

    #[test]
    fn a_legacy_file_converts_itself_when_touched() {
        let dir = scratch();
        let store = dir.join(".md-boss");
        fs::write(&store, r#"{"bookmarks": [{"path": "~/a.md", "line": 3, "title": "Plan"}]}"#).unwrap();
        let read = read_notes(&store);
        write_notes(&store, &read).unwrap();
        let text = fs::read_to_string(&store).unwrap();
        assert!(text.contains("\"notes\""));
        assert!(!text.contains("bookmarks"));
    }
}

File: src/models/notes.ts
// Notes: a marked line, with or without something written about it. Paths are stored
// tilde-abbreviated (`~/dev/notes/plan.md`) so a `.md-boss` file stays readable and
// survives a different home directory. Line numbers are 1-based. One note per (path, line):
// adding over an existing one edits it, which keeps identity stable without UUIDs in a
// file meant to be hand-edited. Bookmarks and comments used to be two types with the same
// two keys; they were the same record - a bookmark is a note nothing has been written on.

import { isUnder, normalizePath } from './paths'

export interface Note {
  path: string
  line: number
  /** Taken from the source line when the note is made, so a list of them is scannable
   *  without opening every file. Comments written before the merge have none. */
  title: string
  /** What you typed. Empty is a plain jump point - what used to be a bookmark. */
  body: string
}

export function note(path: string, line: number, title = '', body = ''): Note {
  return { path, line, title, body }
}

export const noteId = (n: Note) => `${n.path}:${n.line}`
/** What a row leads with. Old comments have no title, and one cannot be invented without
 *  re-reading the file they point at. */
export const noteLabel = (n: Note) => (n.title ? n.title : n.body)
/** What a hover over the line says. A note with no body carries no information beyond its
 *  own existence - its title came off the source line you are already looking at. */
export const noteTooltip = (n: Note) => (n.body ? n.body : `Note on line ${n.line}`)
export const noteIsEmpty = (n: Note) => n.title === '' && n.body.trim() === ''

const byPathThenLine = (a: Note, b: Note) =>
  a.path < b.path ? -1 : a.path > b.path ? 1 : a.line - b.line

// MARK: - The .md-boss file

export interface AnnotationFile {
  notes: Note[]
}

export const emptyFile = (): AnnotationFile => ({ notes: [] })
export const fileIsEmpty = (f: AnnotationFile) => f.notes.length === 0

/** Three shapes fold into one array: `notes`, plus `bookmarks` and `comments` written by
 *  older builds. A line carrying both an old bookmark and an old comment becomes a single
 *  note with a title and a body - the merge, and the one part of it that cannot be undone
 *  once the file is written back. Missing fields read as empty; a malformed file is empty. */
export function parseAnnotationFile(text: string): AnnotationFile {
  let raw: unknown
  try {
    raw = JSON.parse(text)
  } catch {
    return emptyFile()
  }
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) return emptyFile()
  const obj = raw as Record<string, unknown>
  const read = (key: string): Note[] => {
    const list = obj[key]
    if (!Array.isArray(list)) return []
    const out: Note[] = []
    for (const item of list) {
      if (typeof item !== 'object' || item === null) continue
      const { path, line, title, body } = item as Record<string, unknown>
      if (typeof path !== 'string' || typeof line !== 'number') continue
      out.push(note(path, line, typeof title === 'string' ? title : '', typeof body === 'string' ? body : ''))
    }
    return out
  }
  return { notes: fold([...read('notes'), ...read('bookmarks'), ...read('comments')]) }
}

/** Only the current key is written, so a file converts itself the first time anything in
 *  it is touched. Empty fields are left out rather than written as `""`, so a plain jump
 *  point stays a three-key object in a file meant to be read by a person. */
export function serializeAnnotationFile(file: AnnotationFile): string {
  const notes = file.notes.map((n) => {
    const out: Record<string, unknown> = { line: n.line, path: n.path }
    if (n.title) out.title = n.title
    if (n.body) out.body = n.body
    return Object.fromEntries(Object.entries(out).sort(([a], [b]) => (a < b ? -1 : 1)))
  })
  return JSON.stringify({ notes }, null, 2) + '\n'
}

/** One note per (path, line), first non-empty value winning per field. */
export function fold(notes: Note[]): Note[] {
  const merged = new Map<string, Note>()
  for (const n of notes) {
    const id = noteId(n)
    const existing = merged.get(id)
    if (!existing) {
      merged.set(id, { ...n })
      continue
    }
    if (!existing.title) existing.title = n.title
    if (!existing.body) existing.body = n.body
  }
  return [...merged.values()].sort(byPathThenLine)
}

/** Splits the notes on a file that has moved: what stays here, and what has to be
 *  repointed. Null when nothing in this file pointed at `oldPath`, so a store that is not
 *  involved is never rewritten. */
export function repointing(
  file: AnnotationFile,
  oldPath: string,
  newPath: string,
): { kept: AnnotationFile; moved: Note[] } | null {
  if (!file.notes.some((n) => n.path === oldPath)) return null
  const moved: Note[] = []
  const kept: Note[] = []
  for (const n of file.notes) {
    if (n.path === oldPath) moved.push({ ...n, path: newPath })
    else kept.push(n)
  }
  return { kept: { notes: kept }, moved }
}

/** This file without the notes on a document that has gone. Null when it had none. */
export function removingPath(file: AnnotationFile, path: string): AnnotationFile | null {
  if (!file.notes.some((n) => n.path === path)) return null
  return { notes: file.notes.filter((n) => n.path !== path) }
}

// MARK: - Paths

/** The form a note's path is stored in: under `home` it is tilde-abbreviated. */
export function storePath(absolute: string, home: string): string {
  const path = normalizePath(absolute)
  const root = normalizePath(home)
  if (path === root) return '~'
  return path.startsWith(root + '/') ? '~' + path.slice(root.length) : path
}

export function expandPath(stored: string, home: string): string {
  if (stored === '~') return normalizePath(home)
  return stored.startsWith('~/') ? normalizePath(home + stored.slice(1)) : normalizePath(stored)
}

/** First 40 characters of a line, letters, digits and spaces only. Markdown markers,
 *  punctuation and indentation are dropped, so `## The **plan**` suggests "The plan". */
export function suggestedTitle(line: string, limit = 40): string {
  let kept = ''
  for (const ch of line) {
    if (/[\p{L}\p{N}]/u.test(ch)) {
      kept += ch
    } else if (/\s/.test(ch) || ch === '-' || ch === '_') {
      // Collapse runs of separators into a single space.
      if (!kept.endsWith(' ')) kept += ' '
    }
  }
  return Array.from(kept.trim()).slice(0, limit).join('').trim()
}

// MARK: - Scopes

/** The three reaches of the notes pane, nearest first. */
export const NOTE_SCOPES = ['thisFile', 'thisProject', 'allProjects'] as const
export type NoteScope = (typeof NOTE_SCOPES)[number]

export const SCOPE_TITLE: Record<NoteScope, string> = {
  thisFile: 'This file',
  thisProject: 'This project',
  allProjects: 'All projects',
}

/** The open document's notes are the point of the pane, so that scope never folds. */
export const scopeIsCollapsible = (scope: NoteScope) => scope !== 'thisFile'

/** Splits every known note into the three scopes. Pure on purpose. `recentRoots` is the
 *  sidebar's recent list: folders past the twentieth are already unreachable from the
 *  picker, so they contribute nothing here either. Paths are absolute; `home` expands the
 *  stored forms. */
export function partitionNotes(
  all: Note[],
  file: string | null,
  activeRoot: string | null,
  recentRoots: string[],
  home: string,
): Record<NoteScope, Note[]> {
  const currentPath = file === null ? null : storePath(file, home)
  const otherRoots = recentRoots.filter((root) => (activeRoot === null ? true : !isUnder(root, activeRoot)))

  const result: Record<NoteScope, Note[]> = { thisFile: [], thisProject: [], allProjects: [] }
  for (const n of all) {
    const url = expandPath(n.path, home)
    if (currentPath !== null && n.path === currentPath) result.thisFile.push(n)
    else if (activeRoot !== null && isUnder(url, activeRoot)) result.thisProject.push(n)
    else if (otherRoots.some((root) => isUnder(url, root))) result.allProjects.push(n)
  }
  for (const scope of NOTE_SCOPES) result[scope].sort(byPathThenLine)
  return result
}

// MARK: - Across stores

/** One note per (path, line) across *every* store, not just within one file. The copies
 *  fold together field-wise, the same rule the parser and repoint use, so nothing anyone
 *  typed is dropped. `home` names the store a note *should* be in, and a contested one
 *  goes there when it has a copy to spare - otherwise the project's own `.md-boss` could
 *  lose a note to the fallback on nothing but alphabetical luck. A note sitting on its own
 *  is never moved. Stores emptied by the fold still come back in the result, because the
 *  caller has to write them out to finish the repair. */
export function deduplicated(
  stores: Record<string, AnnotationFile>,
  home: (n: Note) => string | null = () => null,
): Record<string, AnnotationFile> {
  // Sorted, so the tie-break does not depend on object ordering.
  const copies = new Map<string, { store: string; note: Note }[]>()
  for (const path of Object.keys(stores).sort()) {
    for (const n of stores[path].notes) {
      const id = noteId(n)
      if (!copies.has(id)) copies.set(id, [])
      copies.get(id)!.push({ store: path, note: n })
    }
  }

  const merged: Record<string, Note[]> = {}
  for (const path of Object.keys(stores)) merged[path] = []
  for (const found of copies.values()) {
    const to = keeper(found, home)
    ;(merged[to] ??= []).push(...found.map((f) => f.note))
  }

  const out: Record<string, AnnotationFile> = {}
  for (const [path, notes] of Object.entries(merged)) out[path] = { notes: fold(notes) }
  return out
}

function keeper(found: { store: string; note: Note }[], home: (n: Note) => string | null): string {
  if (found.length < 2) return found[0].store
  const wanted = home(found[0].note)
  if (wanted === null || !found.some((f) => f.store === wanted)) return found[0].store
  return wanted
}
